### ISSUE TITLE
Add `scale` paramtype2 for nodeboxes, meshes, and plantlike nodes equivalent to `visual_scale`

### ISSUE THREAD

**Issue by Sneglium** [minetest/minetest#15497]
### Problem

Currently nodes must either have a static scale or use one of a few hacky and suboptimal methods to imitate rescaling (excluding having multiple registered node definitions).

### Solutions

The title is mostly self-explanatory, however there are a few possible ways to actually apply the param2 scaling. The ones I can think of are:

- Use the 0-255 value as a kind of percentage of some reasonable scale multiplier, such as `3.0` (e.g. 0 = 0, 127 = ~1.5, 255 = 3).
- Use it as a multiplier for the definition-set `visual_scale`, for example 255 might be 2x the set scale.

These two options have the issue of falling within hardcoded ranges, however, so a third possibility is to use option 1 with the addition of a `visual_scale_max` value in the node definition that determines the multiplier used at param2 = 255, which would allow modders to set it to 1 if they were looking for finer grained control between 0 and 1, or ex. 5 if overall size was more important.

### Alternatives

I can't actually think of any

**Comment by appgurueu** [comment]
> I can't actually think of any

I can. Scaling could be just a special case of a param2-indexed array of transforms per node, [as I suggested in another issue requesting exactly the other two transformations (translation and rotation)](https://github.com/minetest/minetest/issues/15463#issuecomment-2492066466).

This could be much more flexible: No funny hardcoded ranges, no limitations like linearity of scale, no node def or drawtype bloat.

**Comment by Sneglium** [comment]
Interesting concept, but would be really annoying to use for cases where you only need a variable continuous scale modifier. Having to programmatically build a large array of transforms just to have the ability to set the size of a node isn't really ideal for developer time or memory usage relative to the actual effect.

I'd be in favour of a transformation selecting param2 (translation, rotation, scale) alongside a basic scaling-only param2, but having just the array of transforms would be a massive pain.

**Comment by appgurueu** [comment]
> developer time

Developer time is okay, though we probably do need something like a `transform` abstraction for this to be nice  (but we need that anyways), so not just 4x4 matrices. Hypothetical example:

```lua
local transforms = {}
for i = 0, 255 do
    transforms[i] = transform.scale(i / 255)
end
core.register_node("...", {
    transforms = transforms,
    ...
}
```

With a few functional helpers this could even be written as something like `map(function(x) return transform.scale(x / 255) end, range(0, 255))`. Or just make your own `transform_scale_range` for the above first 4 lines of code.

> memory usage

If you calculate the absolute currently possible C++ side upper bound with 256 transforms for *all* 30k node defs (assuming 16 floats per transform), you get ~500 MB. So if this is not used for very many node defs, it is very much unproblematic. (Not to mention: I don't want to require users to supply all 256 transforms; that would only need to be done if necessary. In particular the vast majority of node defs probably don't need this feature and would hence supply just 0 transforms and incur practically no memory usage overhead.)

I doubt that the memory usage ought to be a major concern here. And if it did become one, I would assume that many node defs share the same transform arrays (e.g. in your use case, scale ranges for example), and just save the memory usage by deduplicating those.

> having just the array of transforms would be a massive pain

I don't think "massive" is on the mark. I don't see how a few lines of code (which can be abstracted well and effectively only written once) constitute a massive pain. Even if I was very lazy and just gave you a plain 4x4 matrix to work with, this would *still* be very simple; you'd just have to define a small helper to build a diagonal matrix.

**Comment by Sneglium** [comment]
I still think only having a transform array would be annoying. Mod devs shouldn't need to do any extra work to just change the size of a node, and besides there already exists `degrotate` which would be superseded by your idea so there's precedent before it even exists for having separate ways of changing specific kinds of transformation.

**Comment by appgurueu** [comment]
> I still think only having a transform array would be annoying.

As I already said, we could put the following pure Lua utility in the engine (while personally this feels a bit too specific for me, it doesn't cost much and I don't want to bikeshed excessively so whatever):

```lua
function core.param2_scale_transforms(max_scale)
    local transforms = {}
    for i = 0, 255 do
        transforms[i] = transform.scale(i / 255 * max_scale)
    end
    return transforms
end
```

Then you could do:

```lua
core.register_node("...", {
    paramtype2 = "transform",
    transforms = core.param2_scale_transforms(3),
})
```

At the node def level I think there should ultimately only be pretty much this one general API.

>  there already exists `degrotate` which would be superseded by your idea so there's precedent before it even exists for having separate ways of changing specific kinds of transformation

`degrotate` could be obsoleted indeed, and it is not the only `paramtype2` to be obsoleted. I believe `wallmounted`, `facedir` and `4dir` could be obsoleted all the same. The fact that this single feature essentially lets you pretty straightforwardly implement both past features and present feature requests cleanly IMO is further indication that this is a good abstraction.

In other words: If we could go back in the past and add this one fairly straightforward feature, we wouldn't have had to spend time developing 4 different (and still too specific) features and this feature request (as well as the translation + rotation) one wouldn't exist (or would have been met with a few lines of Lua code, and maybe a trivial convenience PR).

**Comment by Sneglium** [comment]
The `wallmounted` paramtype2 also affects how an `attached_node` behaves, so that would need to be factored in as well. Other than that just having `transform` and `colortransform` would be a big improvement in terms of how dynamic nodes can be.

Currently all those paramtypes allow different maximum palette sizes by just allocating all the bits not needed by the main type as palette indexing. Transformation indexing doesn't need a specific number of bits, so it might be beneficial to have e.g. `color4transform` for 2 bits of palette indexing and 64 transforms, `color8transform` for 3 bits, `color64transform` for 6 bits of palette indexing and 4 transforms and so on. Basic `colortransform` should probably give 4 bits each for color and transformation.

Alternately and a lot more far fetched, the node def could have a generic `states` array each of which can have transform information, color (either from palette index or just a colorspec?) and any other necessary info like which side(s) the node can attach to. A `userstate` or similar param2 could select from these states, effectively giving the modder the ability to define their own paramtype for node appearance. The downside here is more work and probably more breakages, but the values allowed in a 'state' could later be expanded (common example: swapping tiles).

**Comment by sfan5** [comment]
duplicate of #11469
